File: frontend/lib/models/room_model.dart
import 'package:tictac_duel/lib.dart';

class RoomModel {
  const RoomModel({
    required this.id,
    required this.code,
    required this.hostPlayerId,
    required this.occupancy,
    required this.maxRounds,
    required this.currentRound,
    required this.roundStatus,
    required this.theme,
    required this.players,
    required this.turn,
    required this.turnIndex,
    required this.boardSize,
  });

  final String id;
  final String hostPlayerId;
  final String code;
  final int occupancy;
  final int maxRounds;
  final int currentRound;
  final RoomTheme theme;
  final List<PlayerModel> players;
  final PlayerModel? turn;
  final RoundStatus roundStatus;
  final int turnIndex;
  final int boardSize;

  factory RoomModel.fromJson(Map<String, dynamic> json) {
    try {
      return RoomModel(
        id: json['id'] as String,
        code: json['code'] as String,
        hostPlayerId: json['hostPlayerId'] as String,
        occupancy: json['occupancy'] as int,
        maxRounds: json['maxRounds'] as int,
        currentRound: json['currentRound'] as int,
        theme: RoomTheme.values.byName(json['theme'] as String),
        roundStatus: RoundStatus.values.byName(json['roundStatus'] as String),
        players: (json['players'] as List)
            .map(
              (player) => PlayerModel.fromJson(
                Map<String, dynamic>.from(player as Map),
              ),
            )
            .toList(),
        // isPlaying: json['isPlaying'] as bool,
        turn: json['turn'] == null
            ? null
            : PlayerModel.fromJson(
                Map<String, dynamic>.from(json['turn'] as Map),
              ),
        turnIndex: json['turnIndex'] as int,
        boardSize: json['boardSize'] as int,
      );
    } catch (e) {
      rethrow;
    }
  }

  static List<RoomModel> publicRooms = [
    RoomModel(
      id: 'public-room-1',
      hostPlayerId: 'mock-alex',
      code: 'ALEX01',
      occupancy: 1,
      maxRounds: 3,
      currentRound: 0,
      roundStatus: RoundStatus.waiting,
      theme: RoomTheme.classic,
      players: [
        PlayerModel(
          id: 'mock-alex',
          name: 'Alex',
          symbol: PlayerSymbol.x,
          socketId: 'mock-alex',
          points: 0,
          isReady: false,
        ),
      ],
      turn: null,
      turnIndex: 0,
      boardSize: 9,
    ),
    RoomModel(
      id: 'public-room-2',
      hostPlayerId: 'mock-shadow',
      code: 'SHDW01',
      occupancy: 1,
      maxRounds: 5,
      currentRound: 0,
      roundStatus: RoundStatus.waiting,
      theme: RoomTheme.inferno,
      players: [
        PlayerModel(
          id: 'mock-shadow',
          name: 'Shadow',
          symbol: PlayerSymbol.x,
          socketId: 'mock-shadow',
          points: 0,
          isReady: false,
        ),
      ],
      turn: null,
      turnIndex: 0,
      boardSize: 9,
    ),
    RoomModel(
      id: 'public-room-3',
      hostPlayerId: 'mock-nova',
      code: 'NOVA01',
      occupancy: 1,
      maxRounds: 7,
      currentRound: 0,
      roundStatus: RoundStatus.waiting,
      theme: RoomTheme.classic,
      players: [
        PlayerModel(
          id: 'mock-nova',
          name: 'Nova',
          symbol: PlayerSymbol.o,
          socketId: 'mock-nova',
          points: 0,
          isReady: false,
        ),
      ],
      turn: null,
      turnIndex: 0,
      boardSize: 9,
    ),
  ];
}

File: backend/src/models/room_model.js
const mongoose = require('mongoose');

const PLAYER_SYMBOL = Object.freeze({
    X: 'x',
    O: 'o',
});

const ROOM_STATUS = Object.freeze({
    WAITING: 'waiting',
    PLAYING: 'playing',
    RESULT: 'result',
});

const ROOM_THEME = Object.freeze({
    CLASSIC: 'classic',
    CYBERPUNK: 'cyberpunk',
    NEON: 'neon',
});

const playerSchema = new mongoose.Schema(
    {
        name: {
            type: String,
            required: true,
            trim: true,
            minlength: 2,
            maxlength: 20,
        },
        symbol: {
            type: String,
            required: true,
            enum: Object.values(PLAYER_SYMBOL),
        },
        socketId: {
            type: String,
            required: true,
        },
        points: {
            type: Number,
            default: 0,
            min: 0,
        },
        isReady: {
            type: Boolean,
            default: false,
        },
    },
    {
        _id: false,
    },
);

const roomSchema = new mongoose.Schema(
    {
        occupancy: {
            type: Number,
            default: 0,
            min: 0,
            max: 2,
        },
        maxRounds: {
            type: Number,
            default: 5,
            min: 1,
        },
        currentRound: {
            type: Number,
            default: 0,
            min: 0,
        },
        code: {
            type: String,
            required: true,
            unique: true,
            uppercase: true,
            trim: true,
            minlength: 6,
            maxlength: 6,
        },
        theme: {
            type: String,
            required: true,
            enum: Object.values(ROOM_THEME),
        },
        players: {
            type: [playerSchema],
            validate: {
                validator: (players) => players.length <= 2,
                message: 'A room can only have two players.',
            },
        },
        roundStatus: {
            type: String,
            enum: Object.values(ROOM_STATUS),
            default: ROOM_STATUS.WAITING,
        },
        turn: {
            type: playerSchema,
            default: null,
        },
        turnIndex: {
            type: Number,
            default: 0,
            min: 0,
            max: 1,
        },
        boardSize: {
            type: Number,
            default: 9,
            min: 4,
        },
    },
    {
        timestamps: true,
        strict: true,
    },
);

const Room = mongoose.model('Room', roomSchema);

module.exports = {
    Room,
    PLAYER_SYMBOL,
    ROOM_STATUS,
    ROOM_THEME,
};